org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/services/extensions/ExtensionRegistryDoSaveTest.java

```diff
@@ -0,0 +1,98 @@
+/**
+ *  Copyright (c) 2020 Yatao Li.
+ *  All rights reserved. This program and the accompanying materials
+ *  are made available under the terms of the Eclipse Public License v2.0
+ *  which accompanies this distribution, and is available at
+ *  http://www.eclipse.org/legal/epl-v20.html
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ *  Contributors:
+ *  Yatao Li <yatao.li@live.com> - initial API and implementation
+ */
+package org.eclipse.lemminx.services.extensions;
+
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import java.util.function.Predicate;
+
+import org.eclipse.lemminx.XMLAssert.SettingsSaveContext;
+import org.eclipse.lemminx.dom.DOMDocument;
+import org.eclipse.lemminx.services.extensions.IXMLExtension;
+import org.eclipse.lemminx.services.extensions.XMLExtensionsRegistry;
+import org.eclipse.lemminx.services.extensions.save.ISaveContext;
+import org.eclipse.lsp4j.InitializeParams;
+import org.junit.jupiter.api.Test;
+
+import com.google.gson.JsonObject;
+
+/**
+ * Ensures that {@code XMLExtensionsRegistry.doSave()} correctly captures the
+ * initial LS configuration, and the configuration is correctly forwarded to
+ * {@code IXMLExtension.doSave()} for registered extensions.
+ */
+public class ExtensionRegistryDoSaveTest {
+	
+	class RegistryTestExtension implements IXMLExtension {
+		
+		private boolean extensionDoSaveCalled = false;
+		
+		@Override
+		public void start(InitializeParams params, XMLExtensionsRegistry registry) {
+		}
+
+		@Override
+		public void stop(XMLExtensionsRegistry registry) {
+		}
+		
+		@Override
+		public void doSave(ISaveContext context) {
+			extensionDoSaveCalled = true;
+		}
+
+		public boolean isExtensionDoSaveCalled() {
+			return extensionDoSaveCalled;
+		}
+	}
+	
+	private static class SaveFileContext implements ISaveContext {
+		@Override
+		public DOMDocument getDocument(String uri) {
+			return null;
+		}
+
+		@Override
+		public void collectDocumentToValidate(Predicate<DOMDocument> validateDocumentPredicate) {
+		}
+
+		@Override
+		public SaveContextType getType() {
+			return SaveContextType.DOCUMENT;
+		}
+
+		@Override
+		public String getUri() {
+			return null;
+		}
+
+		@Override
+		public Object getSettings() {
+			return null;
+		}
+	}
+
+	@Test
+	public void initialConfiguration() {
+		XMLExtensionsRegistry registry = new XMLExtensionsRegistry();
+		// initial config passed in during server startup.
+		registry.doSave(new SettingsSaveContext(new JsonObject()));
+		// then another event comes in from the client
+		registry.doSave(new SaveFileContext());
+		// extension registration is triggered
+		registry.initializeIfNeeded();
+		RegistryTestExtension extension = new RegistryTestExtension();
+		registry.registerExtension(extension);
+		// examine that the extension properly receives the initial config
+		assertTrue(extension.isExtensionDoSaveCalled());
```

Comment:
I think the test should check that the ISaveContext which is stored is the save context coming from SETTINGS (and not from DOCUMENT which is done before settings).

Reply:
When the context is DOCUMENT it won't make it into the extension DoSave.

Reply:
I think the correct test is:

```java
package org.eclipse.lemminx.services.extensions;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;

import java.util.function.Predicate;

import org.eclipse.lemminx.XMLAssert.SettingsSaveContext;
import org.eclipse.lemminx.dom.DOMDocument;
import org.eclipse.lemminx.services.extensions.save.ISaveContext;
import org.eclipse.lsp4j.InitializeParams;
import org.junit.jupiter.api.Test;

import com.google.gson.JsonObject;

public class XMLExtensionRegistryTest {

	class RegistryTestExtension implements IXMLExtension {

		private ISaveContext context;

		@Override
		public void start(InitializeParams params, XMLExtensionsRegistry registry) {
		}

		@Override
		public void stop(XMLExtensionsRegistry registry) {
		}

		@Override
		public void doSave(ISaveContext context) {
			this.context = context;
		}

		public ISaveContext getContext() {
			return context;
		}
	}

	private static class SaveFileContext implements ISaveContext {
		@Override
		public DOMDocument getDocument(String uri) {
			return null;
		}

		@Override
		public void collectDocumentToValidate(Predicate<DOMDocument> validateDocumentPredicate) {
		}

		@Override
		public SaveContextType getType() {
			return SaveContextType.DOCUMENT;
		}

		@Override
		public String getUri() {
			return null;
		}

		@Override
		public Object getSettings() {
			return null;
		}
	}

	@Test
	public void initialConfiguration() {
		XMLExtensionsRegistry registry = new XMLExtensionsRegistry();
		// initial config passed in during server startup.
		ISaveContext settingsSaveContext  = new SettingsSaveContext(new JsonObject());
		registry.doSave(settingsSaveContext );
		// then another event comes in from the client
		ISaveContext documentSaveContext  = new SaveFileContext();
		registry.doSave(documentSaveContext);
		// extension registration is triggered
		registry.initializeIfNeeded();
		RegistryTestExtension extension = new RegistryTestExtension();
		registry.registerExtension(extension);
		// examine that the extension properly receives the initial config
		assertNotNull(extension.getContext());
		assertEquals(settingsSaveContext.getType(), extension.getContext().getType());
	}
}
```

what do you think about that?

Reply:
Just got back home, no need to RDP into the dev box now :)
Yes I think this looks better.
In af3dfac I updated the test logic.